src/main/java/com/billyyccc/mssqlclient/impl/codec/ExtendedQueryCommandCodec.java

```diff
@@ -0,0 +1,236 @@
+          break;
+        default:
+          throw new UnsupportedOperationException("Unsupported token: " + tokenByte);
+      }
+    }
+  }
+
+  private void sendPrepexecRequest() {
+    ChannelHandlerContext chctx = encoder.chctx;
+
+    ByteBuf packet = chctx.alloc().ioBuffer();
+
+    // packet header
+    packet.writeByte(MessageType.RPC.value());
+    packet.writeByte(MessageStatus.NORMAL.value() | MessageStatus.END_OF_MESSAGE.value());
+    int packetLenIdx = packet.writerIndex();
+    packet.writeShort(0); // set length later
+    packet.writeShort(0x00);
+    packet.writeByte(0x00); // FIXME packet ID
+    packet.writeByte(0x00);
+
+    int start = packet.writerIndex();
+    packet.writeIntLE(0x00); // TotalLength for ALL_HEADERS
+    encodeTransactionDescriptor(packet, 0, 1);
+    // set TotalLength for ALL_HEADERS
+    packet.setIntLE(start, packet.writerIndex() - start);
+
+    /*
+      RPCReqBatch
+     */
+    packet.writeShortLE(0xFFFF);
+    packet.writeShortLE(ProcId.Sp_PrepExec);
+
+    // Option flags
+    packet.writeShortLE(0x0000);
+
+    // Parameter
+
+    // OUT Parameter
+    packet.writeByte(0x00);
+    packet.writeByte(0x01); // By reference
+    packet.writeByte(MSSQLDataTypeId.INTNTYPE_ID);
+    packet.writeByte(0x04);
+    packet.writeByte(0x04);
+    packet.writeIntLE(0x00);
+
+    Tuple params = cmd.params();
+
+    // Param definitions
+    String paramDefinitions = parseParamDefinitions(params);
+    encodeNVarcharParameter(packet, paramDefinitions);
+
+    // SQL text
+    encodeNVarcharParameter(packet, cmd.sql());
+
+    // Param values
+    for (int i = 0; i < params.size(); i++) {
+      encodeParamValue(packet, params.getValue(i));
+    }
+
+    int packetLen = packet.writerIndex() - packetLenIdx + 2;
+    packet.setShort(packetLenIdx, packetLen);
+
+    chctx.writeAndFlush(packet);
+  }
+
+  private String parseParamDefinitions(Tuple params) {
+    StringBuilder stringBuilder = new StringBuilder();
+    for (int i = 0; i < params.size(); i++) {
+      Object param = params.getValue(i);
+      stringBuilder.append("@P").append(i + 1).append(" ");
+      stringBuilder.append(inferenceParamDefinitionByValueType(param));
+      if (i != params.size() - 1) {
+        stringBuilder.append(",");
+      }
+    }
+    return stringBuilder.toString();
+  }
+
+  private void encodeNVarcharParameter(ByteBuf payload, String value) {
+    payload.writeByte(0x00); // name length
+    payload.writeByte(0x00); // status flags
+    payload.writeByte(MSSQLDataTypeId.NVARCHARTYPE_ID);
+    payload.writeShortLE(8000); // maximal length
+    payload.writeByte(0x09);
+    payload.writeByte(0x04);
+    payload.writeByte(0xd0);
+    payload.writeByte(0x00);
+    payload.writeByte(0x34); // Collation for param definitions TODO always this value?
+    writeUnsignedShortLenVarChar(payload, value);
+  }
+
+  private String inferenceParamDefinitionByValueType(Object value) {
```

Comment:
using a Map<Class, String> would be better I think in this case

Reply:
updated in https://github.com/BillyYccc/incubator-vertx-mssql-client/pull/12/commits/681515d889ddd75f516d46dfdff28753dafc4bea.